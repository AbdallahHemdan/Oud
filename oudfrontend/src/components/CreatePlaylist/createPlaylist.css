.createPlaylist {
  background-color: rgba(0, 0, 0, 1);
  min-height: 750px;
  color: #fff;
  text-align: center;
  padding: 37px 0px;
  z-index: 4;
  width: 100%;
}

.closeButton {
  background-color: transparent;
  text-align: center;
  border: transparent;
  padding: 8px 20%;
}
#createPlaylistTitle {
  font-weight: bold;
  margin: 16px 0px;
  line-height: 56px;
  font-size: 48px;
}
#createPlaylistBigField{
    width: 100%;
    padding: 20px 20%;
    background-color: #282828;
    min-height: 148px;
    margin-bottom: 28px;
}
#wrapperId{
    width:100%;
}
#second-wrapperId{
    width:100%;
}
#cardsId{
    width: 100%;
}
#createPlaylistContainer{
    padding: 32px;
    text-align: left;
    font-size: 14px;
    line-height: 20px;
    letter-spacing: .015em;
    font-weight: bold;
    caret-color: #1db954
}
#cretePLaylistName {
  font-size: 48px;
  line-height: 56px;
  width: 100%;
  background-color: transparent;
  color: #fff;
  border: none;
  height: 65px;
  font-weight: bold;
}

#cancelCreation{
    z-index: 2;
    border-style: solid;
    border-width: 2px;
    width: 150px;
    border: 2px solid #f0a91a;
    transition: all 0.4s linear;
    border-radius: 80px !important;
    border-radius: 0.85rem;
    display: inline;
    background-color: transparent;
    color: #f0a91a;
    margin:10px;
}
#cancelCreation:hover {
    border: 2px solid #ffffff;
}
#ceatePlaylistBtn{
    display: inline-block;
    padding: 5px 10px;
    height: 27px;
}
.invisibleButton{
    margin-bottom: 10px;
    background-color: transparent;
    border: none;
    padding:0px;
}
