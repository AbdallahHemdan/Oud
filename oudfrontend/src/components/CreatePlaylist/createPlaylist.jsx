import React, { Component } from "react";
import "./createPlaylist.css";
import axios from "axios";
import getUserId from "../Profile/General/getUserId";
import {base} from "../../config/environment"
import {config} from "../../utils/auth"
import PropTypes from 'prop-types';
import {Auth} from '../../utils/auth';
/**
 * it is an overlay that is used to create a new playlist
 * @class
 * @param {boolean} display true ifthe component is to be visible
 * @property {boolean} display true when the component is visible
 * @property {string} name name of the new playlist
 */
class CreatePlaylist extends Component {
  constructor(props) {
    super(props);
    this.state = {
      display: true,
      name: "",
    };
    this.close = this.close.bind(this)
  }
  /**
   * if the component recieved new props it sets the display property to it
   * @param {object} nextProps new props
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.display !== this.state.display) {
      this.setState({ display: nextProps.display });
    }
  }
  /**
   * updates the state to the contents of the textbox
   * @param {event} e the event of changing the text
   * @returns {void}
   */
  updateName(e) {
    const name = e.target.value;
    this.setState({ name: name });
  }
  /**
   * creates a playlist object and sends to the database
   * @returns {void}
   */
  createPlaylist() {
    let playlist = {
      name: this.state.name,
      public: true,
      description: "my playlist",
    };
    let id = Auth();
    axios
      .post(`${base}/users/${id}/playlists`, playlist, config)
      .then(function(response) {
        console.log(response);
      })
      .catch(function (error) {
      });
      this.close()
  }
  /**
   * closes the window by making state.display false
   * @returns {void}
   */
  close() {
    if(this.props.close !== undefined)
      {this.props.close()}
    this.setState({ display: false });
    
  }
  render() {
    if(this.state.display)
    {return (
      <div
        className="createPlaylist"
        
        data-testid='createPlaylist'
      >
        <button onClick={this.close.bind(this)} className="closeButton" data-testid='closeButton'>
          <svg width="32" height="32" xmlns="http://www.w3.org/2000/svg" >
            <title>close</title>
            <path
              d="M31.098 29.794L16.955 15.65 31.097 1.51 29.683.093 15.54 14.237 1.4.094-.016 1.508 14.126 15.65-.016 29.795l1.414 1.414L15.54 17.065l14.144 14.143"
              fill="#fff"
              fillRule="evenodd"
            ></path>
          </svg>
        </button>
        <h1 id="createPlaylistTitle" data-testid='title'>Create new playlist</h1>

        <div id="createPlaylistBigField">
          <div id="createPlaylistContainer" data-testid='inputContainer'>
            <p className="gray-text" data-testid='inputHeader'>Playlist Name</p>
            <input
              data-testid='input'
              id="cretePLaylistName"
              type="text"
              onChange={this.updateName.bind(this)}
              placeholder="New Playlist"
            />
          </div>
        </div>
        <button data-testid='cancelButton' id="cancelCreation" onClick={()=>this.close()}>
          CANCEL
        </button>
        <button data-testid='CreateButton'  id="cancelCreation" onClick={this.createPlaylist.bind(this)}>
          CREATE
        </button>
      </div>
    );
  }
  else{
    return(<div data-testid='empty'></div>)
  }
  }
}
CreatePlaylist.propTypes ={
  display:PropTypes.bool
}
export default CreatePlaylist;
