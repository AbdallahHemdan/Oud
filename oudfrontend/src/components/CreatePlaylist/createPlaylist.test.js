import React from 'react'
import CreatePlaylist from './createPlaylist';
import Enzyme, { shallow } from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer';
import checkPropTypes from 'check-prop-types'

Enzyme.configure({ adapter: new EnzymeAdapter() });


const setup = () => {
    return shallow(<CreatePlaylist />);
}

const findByTestAttr = (wrapper, val) => {
    return wrapper.find(`[data-testid="${val}"]`);
}
/*
Rendering tests
*/

describe('CreatePlaylist Component', () => {
    describe('testing that CreatePlaylist renders Correctly when display = true', () => {
        let component;
        beforeEach(() => {
            component = setup();
            component.setState({ display: true })
        })
        it('renders createPlaylist component', () => {

            const wrapper = findByTestAttr(component, "createPlaylist");
            expect(wrapper.length).toBe(1);
        });
        it('not render empty component', () => {

            const wrapper = findByTestAttr(component, "empty");
            expect(wrapper.length).toBe(0);
        });
        it('renders closeButton component', () => {
            const wrapper = findByTestAttr(component, "closeButton");
            expect(wrapper.length).toBe(1);
        });
        it('renders title component', () => {
            const wrapper = findByTestAttr(component, "title");
            expect(wrapper.length).toBe(1);
        });
        it('renders inputContainer component', () => {
            const wrapper = findByTestAttr(component, "inputContainer");
            expect(wrapper.length).toBe(1);
        });
        it('renders inputHeader component', () => {
            const wrapper = findByTestAttr(component, "inputHeader");
            expect(wrapper.length).toBe(1);
        });
        it('renders input component', () => {
            const wrapper = findByTestAttr(component, "input");
            expect(wrapper.length).toBe(1);
        });
        it('renders CancelButton component', () => {
            const wrapper = findByTestAttr(component, "cancelButton");
            expect(wrapper.length).toBe(1);
        });
        it('renders CreateButton component', () => {
            const wrapper = findByTestAttr(component, "CreateButton");
            expect(wrapper.length).toBe(1);
        });

    });
    describe('testing that CreatePlaylist renders Correctly when display = false', () => {
        let component;
        beforeEach(() => {
            component = setup();
            component.setState({ display: false })
        })
        it('renders createPlaylist component', () => {

            const wrapper = findByTestAttr(component, "createPlaylist");
            expect(wrapper.length).toBe(0);
        });
        it('renders empty component', () => {

            const wrapper = findByTestAttr(component, "empty");
            expect(wrapper.length).toBe(1);
        });
        it('renders closeButton component', () => {
            const wrapper = findByTestAttr(component, "closeButton");
            expect(wrapper.length).toBe(0);
        });
        it('renders title component', () => {
            const wrapper = findByTestAttr(component, "title");
            expect(wrapper.length).toBe(0);
        });
        it('renders inputContainer component', () => {
            const wrapper = findByTestAttr(component, "inputContainer");
            expect(wrapper.length).toBe(0);
        });
        it('renders inputHeader component', () => {
            const wrapper = findByTestAttr(component, "inputHeader");
            expect(wrapper.length).toBe(0);
        });
        it('renders input component', () => {
            const wrapper = findByTestAttr(component, "input");
            expect(wrapper.length).toBe(0);
        });
        it('renders CancelButton component', () => {
            const wrapper = findByTestAttr(component, "cancelButton");
            expect(wrapper.length).toBe(0);
        });
        it('renders CreateButton component', () => {
            const wrapper = findByTestAttr(component, "CreateButton");
            expect(wrapper.length).toBe(0);
        });

    });
    describe('testing that CreatePlaylist texts', () => {
        let component;
        beforeEach(() => {
            component = setup();
        })
        it('testing the title text', () => {
            const wrapper = findByTestAttr(component, "title");
            expect(wrapper.text()).toBe('Create new playlist');
        });
        it('testing input header text', () => {
            const wrapper = findByTestAttr(component, "inputHeader");
            expect(wrapper.text()).toBe('Playlist Name');
        });
        it('testing the cancelButton text', () => {
            const wrapper = findByTestAttr(component, "cancelButton");
            expect(wrapper.text()).toBe('CANCEL');
        });
        it('testing input createButton text', () => {
            const wrapper = findByTestAttr(component, "CreateButton");
            expect(wrapper.text()).toBe('CREATE');
        });
    });
    describe('testing actions', () => {
        let component;
        beforeEach(() => {
            component = setup();
            component.setState({ display: true })

        })
        it('close Button', () => {
            const wrapper = findByTestAttr(component, "closeButton");
            expect(component.state().display).toBeTruthy();
            wrapper.simulate('click')
            expect(component.state().display).toBeFalsy();
        });
        it('cancel Button', () => {
            const wrapper = findByTestAttr(component, "cancelButton");
            expect(component.state().display).toBeTruthy();
            wrapper.simulate('click')
            expect(component.state().display).toBeFalsy();
        });
        it('create Button', () => {
            const wrapper = findByTestAttr(component, "CreateButton");
            expect(component.state().display).toBeTruthy();
            wrapper.simulate('click')
            expect(component.state().display).toBeFalsy();
        });
        it('create Button', () => {
            const wrapper = findByTestAttr(component, "CreateButton");
            expect(component.state().display).toBeTruthy();
            wrapper.simulate('click')
            expect(component.state().display).toBeFalsy();
        });
        it('text change', () => {
            const wrapper = findByTestAttr(component, "input");
            expect(component.state().name).toBe("");
            wrapper.simulate('change', { target: { value: 'Hello' } })
            expect(component.state().name).toBe("Hello");
        });
    })
    describe('testing proptypes', () => {
        let component;
        beforeEach(() => {
            component = setup();
        })
        it('testing with true props', () => {
            const result = checkPropTypes(CreatePlaylist.propTypes, { display: true }, 'prop', CreatePlaylist.name);
            expect(result).toBeUndefined();
        });
        it('testing with false props', () => {
            const result = checkPropTypes(CreatePlaylist.propTypes, { display: 55 }, 'prop', CreatePlaylist.name);
            expect(result).toBeDefined();
        });
        it('testing with false props', () => {
            const result = checkPropTypes(CreatePlaylist.propTypes, { display: {} }, 'prop', CreatePlaylist.name);
            expect(result).toBeDefined();
        });
        it('testing with false props', () => {
            const result = checkPropTypes(CreatePlaylist.propTypes, { display: "gg" }, 'prop', CreatePlaylist.name);
            expect(result).toBeDefined();
        });
    });
    describe('componentWillReceiveProps()', () => {
        it('change the display state', () => {
            const component = shallow(<CreatePlaylist display={true} />);
            expect(component.state().display).toBe(true);
            component.setProps({ display: false });
            expect(component.state().display).toBe(false);
        })
        it('does not change the display state', () => {
            const component = shallow(<CreatePlaylist display={true} />);
            expect(component.state().display).toBe(true);
            component.setProps({ display: true });
            expect(component.state().display).toBe(true);
        })
    });
    describe('snapshot test', () => {
        it(' correctly', () => {
            const tree = renderer
                .create(<CreatePlaylist />)
                .toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

});