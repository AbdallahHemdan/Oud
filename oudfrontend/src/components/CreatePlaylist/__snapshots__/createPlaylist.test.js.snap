// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePlaylist Component snapshot test  correctly 1`] = `
<div
  className="createPlaylist"
  data-testid="createPlaylist"
>
  <button
    className="closeButton"
    data-testid="closeButton"
    onClick={[Function]}
  >
    <svg
      height="32"
      width="32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        close
      </title>
      <path
        d="M31.098 29.794L16.955 15.65 31.097 1.51 29.683.093 15.54 14.237 1.4.094-.016 1.508 14.126 15.65-.016 29.795l1.414 1.414L15.54 17.065l14.144 14.143"
        fill="#fff"
        fillRule="evenodd"
      />
    </svg>
  </button>
  <h1
    data-testid="title"
    id="createPlaylistTitle"
  >
    Create new playlist
  </h1>
  <div
    id="createPlaylistBigField"
  >
    <div
      data-testid="inputContainer"
      id="createPlaylistContainer"
    >
      <p
        className="gray-text"
        data-testid="inputHeader"
      >
        Playlist Name
      </p>
      <input
        data-testid="input"
        id="cretePLaylistName"
        onChange={[Function]}
        placeholder="New Playlist"
        type="text"
      />
    </div>
  </div>
  <button
    data-testid="cancelButton"
    id="cancelCreation"
    onClick={[Function]}
  >
    CANCEL
  </button>
  <button
    data-testid="CreateButton"
    id="cancelCreation"
    onClick={[Function]}
  >
    CREATE
  </button>
</div>
`;
